src/Units/Unit.java

```diff
@@ -201,62 +245,96 @@ public int getCaptureProgress()
+          // Evaluate attack options.
+          {
+            boolean moved = !moveLocation.equals(x, y);
+            ArrayList<GameAction> attackOptions = new ArrayList<GameAction>();
+            for( Weapon wpn : weapons )
             {
-              for( int h = 0; h < map.mapHeight; h++ )
+              // Evaluate this weapon for targets if it has ammo, and if either the weapon
+              // is mobile or we don't care if it's mobile (because we aren't moving).
+              if( wpn.ammo > 0 && (!moved || wpn.model.canFireAfterMoving) )
               {
-                if( map.getLocation(w, h).isHighlightSet() )
+                ArrayList<XYCoord> locations = Utils.findTargetsInRange(map, CO, moveLocation, wpn);
+
+                for( XYCoord loc : locations )
                 {
-                  actions.add(GameAction.ActionType.ATTACK);
-                  found = true;
-                  break; // We just need one target to know attacking is possible.
+                  attackOptions.add(new GameAction.AttackAction(map, this, movePath, loc));
                 }
               }
-              if( found )
-                break;
+            } // ~Weapon loop
+
+            // Only add this action set if we actually have a target
+            if( !attackOptions.isEmpty() )
+            {
+              // Bundle our attack options into an action set and add it to our return collection.
+              actionSet.add(new GameActionSet(attackOptions));
             }
-            map.clearAllHighlights();
+          } // ~attack options
             break;
           case CAPTURE:
-            if( map.getLocation(xLoc, yLoc).getOwner() != CO && map.getLocation(xLoc, yLoc).isCaptureable() )
+            if( map.getLocation(moveLocation).getOwner() != CO && map.getLocation(moveLocation).isCaptureable() )
             {
-              actions.add(GameAction.ActionType.CAPTURE);
+              actionSet.add(new GameActionSet(new GameAction.CaptureAction(map, this, movePath), false));
             }
             break;
           case WAIT:
-            actions.add(GameAction.ActionType.WAIT);
+            actionSet.add(new GameActionSet(new GameAction.WaitAction(map, this, movePath), false));
             break;
           case LOAD:
-            // Don't add - there's no unit there to board.
+            // We only get to here if there is no unit at the end of the move path, which means there is
+            //   no transport in this space to board. LOAD actions are handled down below.
             break;
           case UNLOAD:
             if( heldUnits.size() > 0 )
             {
-              actions.add(GameAction.ActionType.UNLOAD);
+              ArrayList<GameAction> unloadActions = new ArrayList<GameAction>();
+
+              // TODO: This could get messy real quick for transports with more cargo space. Figure out a
+              //       better way to handle this case.
```

Comment:
Is having an instance of each unit/point pair a problem?
You could swap to iterating through each of the 4 spaces first, and just checking if any unit can unload there.
Maybe the best way would be to just make a new GameActionSet for each unit, and display the unit name instead of "unload" in the action menu.
This is particularly appealing because it reduces "clicks" and because units that can unload typically don't have many other actions available.